include/nbl/asset/interchange/IAssetLoader.h

```diff
@@ -289,6 +289,42 @@ class IAssetLoader : public virtual core::IReferenceCounted
 	SAssetBundle interm_getAssetInHierarchy(IAssetManager* _mgr, const std::string& _filename, const IAssetLoader::SAssetLoadParams& _params, uint32_t _hierarchyLevel, IAssetLoader::IAssetLoaderOverride* _override);
 	SAssetBundle interm_getAssetInHierarchy(IAssetManager* _mgr, io::IReadFile* _file, const std::string& _supposedFilename, const IAssetLoader::SAssetLoadParams& _params, uint32_t _hierarchyLevel);
 	SAssetBundle interm_getAssetInHierarchy(IAssetManager* _mgr, const std::string& _filename, const IAssetLoader::SAssetLoadParams& _params, uint32_t _hierarchyLevel);
+	//! ECF_DUPLICATE_* flags are ignored (since there's no point in **restoring** if we're not even looking into the cache) -- only ECF_DONT_CACHE_* are honored
+	//! `Args` ought to be same parameter types as interm_getAssetInHierarchy(...) overloads offer
+	template <typename ...Args>
+	SAssetBundle interm_getRestoredAssetInHierarchy(uint32_t _restoreLevels, Args&&... args)
+	{
+		if (_restoreLevels == 0u)
+			return interm_getAssetInHierarchy(std::forward<Args>(args)...);
+
+		auto any_dummy = [](const SAssetBundle& b) {
+			for (const auto& a : b.getContents())
+				if (a->isADummyObjectForCache())
+					return true;
+		};
+
+		auto bundle = interm_getAssetInHierarchy_find(std::forward<Args>(args)...);
+
+		if (bundle.getContents().empty() || !any_dummy(bundle))
+			return bundle;
+		auto reloadedBundle = interm_getAssetInHierarchy_reload(_restoreLevels, std::forward<Args>(args)...);
+
+		assert(bundle.getContents().size() == reloadedBundle.getContents().size());
+
+		const uint32_t count = bundle.getMutableContents().size();
+		auto* dummies = bundle.getMutableContents().begin();
+		auto* reloaded = reloadedBundle.getMutableContents().begin();
+		for (uint32_t i = 0u; i < count; ++i)
+			if (dummies[i]->isADummyObjectForCache() && !dummies[i]->canBeRestoredFrom(reloaded[i].get()))
+				return {}; // return empty bundle
+
+		for (uint32_t i = 0u; i < count; ++i)
+			if (dummies[i]->isADummyObjectForCache())
+				dummies[i]->restoreFromDummy(reloaded[i].get(), _restoreLevels);
```

Comment:
this part should be inside the IAssetLoaderOverride and it should be the one choosing what in the first bundle gets restored from what in the second bundle

Reply:
```cpp
chosenAsset = override->handleRestore(std::move(chosenAsset), initialBundle, reloadedBundle)
if (chosenAsset->isAnyDependencyDummy(restoreLevel-1))
   return nullptr;
return chosenAsset;
```